[WiP] lifecycle support

norro opened the pull request:
Adds lifecycle support to rclc. Introduces a new struct for a `lifecycle_node`, bundling an rcl node with the ROS 2 default state machine, as well as convenience functions to register callbacks and transition through the state machine.
See [https://github.com/micro-ROS/rclc/blob/feature/lifecycle/rclc_examples/src/lifecycle_node.c](rclc_examples/src/lifecycle_node.c) for an example.

I decided to go for a separate package in the same way it is done with `rcl_lifecycle` and `rclcpp_lifecycle`.

See https://github.com/micro-ROS/rclc/issues/37

norro committed:
Adds rcl lifecycle convenience functions

* Introduces lifecycle_struct that bundles and rcl node with the ROS 2
default state machine
* Introduces convenience methods to register callbacks on lifecycle
state changes

https://github.com/micro-ROS/rclc/issues/37

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Adds lifecycle example to rclc_examples

Example creates a lifecycle_node, registers callbacks for state changes,
and then programmatically runs through the state machine.

https://github.com/micro-ROS/rclc/issues/37

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Started linting and testing

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

pablogs9 commented:
Hello @norro, we were about to migrate ROS 2 parameters from RCL to RCLC. Currently, they are in this [RCL branch](https://github.com/micro-ROS/rcl/tree/feature/c_parameters_uros). I was wondering if the lifecycle uses parameters or it is independent?

Thanks!

norro committed:
All tests passing

Adapted API, lifecycle node now passed asparameter for initialization, c
header included as "extern C".

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Further tests: lifeycycle node transitions

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Unit test for lifecycle change hooks/callbacks

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro commented:
@JanStaschulat I appreciate all feedback you have, concerning API as well as c coding style.

norro committed:
Added documentation for rclc_lifecycle

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

ralph-lange commented on README.md:
```diff
@@ -1,10 +1,11 @@
 
 - [rclc](rclc/) provides the mentioned convenience functions for creating instances of publishers, subscriptions, nodes, etc. with the corresponding [rcl types](https://github.com/ros2/rcl/tree/master/rcl/include/rcl). Furthermore, it provides the rclc Executor for C, analogously to rclcpp's [Executor class](https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/executor.hpp) for C++. A ...
-- [rclc_examples](rclc_examples/) provides small examples for the use of the convenience functions and the rclc Executor.
+- [rclc_lifecycle](rclc_lifecycle/) introduces an rclc Lifecycle Node, bundling an rcl Node and the ROS 2 state machine.
```
```suggestion
- [rclc_lifecycle](rclc_lifecycle/) introduces an rclc Lifecycle Node, bundling an rcl Node and the [lifecycle state machine](http://design.ros2.org/articles/node_lifecycle.html) from the [rcl_lifecycle package](https://github.com/ros2/rcl/tree/master/rcl_lifecycle).
```

ralph-lange commented on rclc_examples/CMakeLists.txt:
```diff
@@ -33,11 +36,14 @@ ament_target_dependencies(example_executor_convenience rcl rclc std_msgs)
 add_executable(example_executor_trigger src/example_executor_trigger.c)
 ament_target_dependencies(example_executor_trigger rcl rclc std_msgs)
 
+add_executable(lifecycle_node src/lifecycle_node.c)
```
```suggestion
add_executable(example_lifecycle_node src/lifecycle_node.c)
```
Propose to use a consistent naming.

ralph-lange commented on rclc_examples/CMakeLists.txt:
```diff
@@ -33,11 +36,14 @@ ament_target_dependencies(example_executor_convenience rcl rclc std_msgs)
 ament_target_dependencies(example_executor_trigger rcl rclc std_msgs)
 
+add_executable(lifecycle_node src/lifecycle_node.c)
+ament_target_dependencies(lifecycle_node rcutils rcl rcl_lifecycle rclc rclc_lifecycle lifecycle_msgs)
```
```suggestion
ament_target_dependencies(example_lifecycle_node rcutils rcl rcl_lifecycle rclc rclc_lifecycle lifecycle_msgs)
```

ralph-lange commented on rclc_examples/CMakeLists.txt:
```diff
@@ -33,11 +36,14 @@ ament_target_dependencies(example_executor_convenience rcl rclc std_msgs)
   example_executor
   example_executor_convenience
   example_executor_trigger
+  lifecycle_node
```
```suggestion
  example_lifecycle_node
```

ralph-lange commented on rclc_examples/package.xml:
```diff
@@ -3,22 +3,27 @@
-  <exec_depend>rclc</exec_depend>
-  <exec_depend>std_msgs</exec_depend>
+  <build_depend>rcl</build_depend>
+  <build_depend>rclc</build_depend>
```
Be careful with build_depend vs exec_depend. Now there are two build_depend for rcl and two for rclc.

norro replied:
whoops, good find!

ralph-lange commented on rclc_examples/src/lifecycle_node.c:
```diff
@@ -0,0 +1,143 @@
+#include <stdio.h>
```
Please add standard copyright header, cf. for example https://github.com/ros2/rclc/blob/master/rclc/src/rclc/node.c

ralph-lange commented on rclc_examples/src/lifecycle_node.c:
```diff
@@ -0,0 +1,143 @@
+#include <lifecycle_msgs/msg/transition_description.h>
+#include <lifecycle_msgs/msg/transition_event.h>
+#include <lifecycle_msgs/srv/change_state.h>
+#include <lifecycle_msgs/srv/get_state.h>
```
Sort alphabetically within a group a includes, i.e. move this line down by two lines.

ralph-lange commented on rclc_examples/src/lifecycle_node.c:
```diff
@@ -0,0 +1,143 @@
+  return RCL_RET_OK;
+}
+
+rcl_ret_t foobar() {
```
Is there a specific reason for the name foobar instead of on_deactivate.

norro replied:
Just to indicate that the callback name is arbitrary and not necessarily `on_{statename}` as it is in `rclcpp` for example (inheritance).

ralph-lange commented on rclc_examples/src/lifecycle_node.c:
```diff
@@ -0,0 +1,143 @@
+  }
+
+  // create context
+  rc = rcl_init(argc, argv, &init_options, &context);
```
Can `rclc_support_init` be used to save some code lines here and above?

ralph-lange commented on rclc_examples/src/lifecycle_node.c:
```diff
@@ -0,0 +1,143 @@
+  printf("creating rcl node...\n");
+  rcl_node_t my_node = rcl_get_zero_initialized_node();
+  rcl_node_options_t node_ops = rcl_node_get_default_options();
+  rc = rcl_node_init(&my_node, "lifecycle_node", "rclc", &context, &node_ops);
```
Can `rclc_node_init_default` be used to save some code lines here?

ralph-lange commented on rclc_examples/src/lifecycle_node.c:
```diff
@@ -0,0 +1,143 @@
+    
+  // register callbacks
+  rclc_lifecycle_register_on_configure(&lifecycle_node, &on_configure);
+  rclc_lifecycle_register_on_deactivate(&lifecycle_node, &foobar);
```
If you want to emphasize that the use can decide about the function name, I propose to use `my_on_deactivate` instead.

norro replied:
good idea.

ralph-lange commented on rclc_lifecycle/CMakeLists.txt:
```diff
@@ -0,0 +1,75 @@
+  src/rclc_lifecycle/rclc_lifecycle.c
+)
+if("${rcl_VERSION}" VERSION_LESS "1.0.0")
+  target_sources(${PROJECT_NAME} PRIVATE src/rclc/rcl_wait_set_is_valid_backport.c)
```
Is this needed here? The rclc_lifecycle package is not using wait_set_is_valid directly, or? If this is needed to build against rclc, we should think of a better mechanism in the rclc package.

norro replied:
It was not required, thanks for the hint!

ralph-lange commented on rclc_lifecycle/README.md:
```diff
@@ -0,0 +1,84 @@
+
+## Overview
+
+The rclc_lifecycle package is a [ROS 2](http://www.ros2.org/) package, which provides convenience functions to create ROS Client Library(RCL) data types and an RCLC Lifecycle Node in the C programming language.
```
```suggestion
The rclc_lifecycle package is a [ROS 2](http://www.ros2.org/) package and provides convenience functions to bundle a ROS Client Library (RCL) node with the ROS 2 Node Lifecycle state machine in the C programming language, similar to the [rclcpp Lifecycle Node](https://github.com/ros2/rclcpp/blob/master/rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp) for C++.
```

ralph-lange commented on rclc_lifecycle/README.md:
```diff
@@ -0,0 +1,84 @@
+## Overview
+
+The rclc_lifecycle package is a [ROS 2](http://www.ros2.org/) package, which provides convenience functions to create ROS Client Library(RCL) data types and an RCLC Lifecycle Node in the C programming language.
+The convenience functions are a thin API layer on top of RCL-layer to bunde an rcl node with the ROS 2 default state machine similar to the rclcpp Lifecycle Node.#
```
```suggestion

```

ralph-lange commented on rclc_lifecycle/README.md:
```diff
@@ -0,0 +1,84 @@
+The rclc_lifecycle package is a [ROS 2](http://www.ros2.org/) package, which provides convenience functions to create ROS Client Library(RCL) data types and an RCLC Lifecycle Node in the C programming language.
+The convenience functions are a thin API layer on top of RCL-layer to bunde an rcl node with the ROS 2 default state machine similar to the rclcpp Lifecycle Node.#
+
+The rclc Lifecycle Node is a ROS 2 Lifecycle Node implemented based on and for the rcl API, for applications written in the C language.
```
```suggestion

```

ralph-lange commented on rclc_lifecycle/README.md:
```diff
@@ -0,0 +1,84 @@
+
+### Initialization
+
+Creation of a lifecycle node as a bundle of an rcl node and the rcl state machine.
```
```suggestion
Creation of a lifecycle node as a bundle of an rcl node and the rcl Node Lifecycle state machine.
```

ralph-lange commented on rclc_lifecycle/README.md:
```diff
@@ -0,0 +1,84 @@
+
+// rcl node and node options
+rcl_node_t my_node = rcl_get_zero_initialized_node();
+rcl_node_options_t node_ops = rcl_node_get_default_options();
```
Use `rclc_node_init_default` if possible.

ralph-lange commented on rclc_lifecycle/README.md:
```diff
@@ -0,0 +1,84 @@
+
+### Running
+
+Change states of the lifecycle node, e.g.
```
Normally the states are changed from outside (from launch, other nodes or command line tools), except for switching to error-processing, or? I propose to add a comment on this here.

ralph-lange commented on rclc_lifecycle/include/rclc_lifecycle/rclc_lifecycle.h:
```diff
@@ -0,0 +1,88 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef RCLC_LIFECYCLE__RCLC_LIFECYCLE_H_
```
Minor: Add empty line between copyright header and header guard.

ralph-lange commented on rclc_lifecycle/include/rclc_lifecycle/rclc_lifecycle.h:
```diff
@@ -0,0 +1,88 @@
+
+typedef struct rclc_lifecycle_callback_map_t
+{
+  bool goal_states[100];  // highest transition id: 99
```
Introduce define for the 100.

ralph-lange commented on rclc_lifecycle/src/rclc_lifecycle/rclc_lifecycle.c:
```diff
@@ -0,0 +1,256 @@
+  rclc_lifecycle_node_t * lifecycle_node,
+  unsigned int goal_state,
+  int (* cb)(void)
+)
```
Move `)` at the end of the previous line as in all other functions.

ralph-lange commented on rclc_lifecycle/src/rclc_lifecycle/rclc_lifecycle.c:
```diff
@@ -0,0 +1,256 @@
+
+  // Check for callbacks for this transition
+  if (rclc_lifecycle_execute_callback(lifecycle_node, transition_id) == RCL_RET_OK) {
+    // successfull, so transition do according success transition
```
```suggestion
    // successful, so transition do according success transition
```

ralph-lange reviewed (changes requested):
Nice work. I have minor comments only. An additional idea: What about extending or adding an example (or at least unit-test) on the use of the error-processing transition, as this is the only internally-triggered transition.

norro committed:
Incorporates feedback by rlange

* Renaming of example + executable
* Minor changes in rcl_init => no more warnings

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Incorporates further feeback by rlange

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>

norro committed:
Updates documentation and file headers

Signed-off-by: Arne Nordmann <arne.nordmann@de.bosch.com>
